### PR TITLE
Integrated gradients for transformers models

### PR DESCRIPTION
**Pull request by gipster**
# Integrated gradients for transformers models
Transformers models are widely used for tasks such as NLP. The aim of this pull request is to extend the integrated gradients method in [alibi](https://github.com/SeldonIO/alibi/blob/master/alibi/explainers/integrated_gradients.py) to the pretrained tensorflow models provided by the [transformers](https://github.com/huggingface/transformers) library.

## Usage of alibi with transformers library
With a minimal change in the present alibi code, it is possible to obtain the integrated gradients attributions for the input embeddings of a transformer model. A minimal example is provided below  (see issues #362 and #377 )

```python
import tensorflow as tf
import numpy as np
from transformers import TFBertModel
from alibi.explainers import IntegratedGradients

# load the pretrained bert model
modelBert = TFBertModel.from_pretrained('bert-base-uncased')

# define the custom model using bert as a first layer
max_len=512
input_ids_in = tf.keras.layers.Input(shape=(max_len,), 
                                     name='input_tokens', 
                                     dtype=tf.int32)
attention_masks_in = tf.keras.layers.Input(shape=(max_len,), 
                                           name='attention_mask_tokens', 
                                           dtype=tf.int32)
X = modelBert([input_ids_in, attention_masks_in])[0]
# adding a flatten layer as a mock layer 
X = tf.keras.layers.Flatten()(X)
X = tf.keras.layers.Dense(1)(X)
frozenModelOut = tf.keras.Model(inputs=[input_ids_in, attention_masks_in], 
                                outputs=X)
frozenModelOut.layers[2].trainable = False

# extract the input embedding layer from bert model
iel = frozenModelOut.layers[2].bert.embeddings

# define random inputs for testing
x = [np.random.randint(1000, size=(2, max_len)).astype('int32'),
    np.ones((2, max_len)).astype('int32')]

# calculate integrated gradients
ig  = IntegratedGradients(frozenModelOut, layer=iel, internal_batch_size=5)
expl = ig.explain(x)                          
```

The only change required for the examples above to run is to allow the assignment of `layer_num = None` in the metadata output. This is necessary because the extracted layer doesn't have an explicit index in the model. This change is included in the present version of the PR.

## Inner and output transformers layer
In order to use the method above to calculate the attributions for the inner and the output layer of the transformer, it is necessary to have a method to extract those layers from the pretrained bert model, in a way that is analogous to the one described above for the input embeddings. This is currently under investigation. Preliminary tests show that

* The pretrained models in [transformers](https://github.com/huggingface/transformers) do not seem to have a method to extract the inner layers analogous to `get_input_embeddings()`.

* The method `get_output_embeddings()` returns None. 

The goal is to find a systematic way to extract the inner layers from a tensorflow [transformers](https://github.com/huggingface/transformers) model. 

## Work around: wrapping bert in a tf.keras layer
A work around to the problem above is to wrap the bert model into a keras or tf.keras layer when part of a bigger model. 
In this case, the use of the `IntegratedGradients` class is straightforward.
Let's assume we want to use the output of the bert model or the output of one of its internal layers as a starting point for a custom model such as the one considered in the example above. The integrated gradients attributions with respect to the output of the bert layer can be calculated as

```python
import tensorflow as tf
import numpy as np
from transformers import TFBertModel, BertConfig
from alibi.explainers import IntegratedGradients

# load the pretrained bert model and configure it to return all hidden states
config = BertConfig.from_pretrained("bert-base-uncased", output_hidden_states=True)
modelBert = TFBertModel.from_pretrained('bert-base-uncased', config=config)

# wrap the model into a tf.keras layer
class BertLayer(tf.keras.layers.Layer):
    def __init__(self, modelBert, **kwargs):
        super(BertLayer, self).__init__()
        self.modelBert = modelBert

    def call(self, inputs, output_nb=5):
        return self.modelBert(inputs)[2][output_nb]
bertLayer = BertLayer(modelBert, trainable=False)

# define custom model
max_len=512
input_ids_in = tf.keras.layers.Input(shape=(max_len,), 
                                     name='input_tokens', 
                                     dtype=tf.int32)
attention_masks_in = tf.keras.layers.Input(shape=(max_len,), 
                                           name='attention_mask_tokens', 
                                           dtype=tf.int32)
X = bertLayer([input_ids_in, attention_masks_in])
# adding a flatten layer as a mock layer 
X = tf.keras.layers.Flatten()(X)
X = tf.keras.layers.Dense(1)(X)
frozenModelOut = tf.keras.Model(inputs=[input_ids_in, attention_masks_in], 
                                outputs=X)

# extract the input embedding layer from bert model
el = frozenModelOut.layers[2]

# define random inputs for testing
x = [np.random.randint(1000, size=(2, max_len)).astype('int32'),
    np.ones((2, max_len)).astype('int32')]

# calculate integrated gradients
ig  = IntegratedGradients(frozenModelOut, layer=el, internal_batch_size=5)
expl = ig.explain(x)                          
```

Note that the pretrained bert model must be configured in order to return all hidden states, and the layer must specify in advance which hidden state must be return. This work around is not ideal, but it allows to calculate the attributions of the inner and output embeddings in a limited number of use cases. 

A more permanent solution is under investigation.

### CODE REVIEW

**Comment by jklaise** [comment]
I think this is how we can proceed.

1. Work on a real-world example with a transformer model explaining some real text data and showcase the 2 ways described above how this can be achieved.
2. The proposed handling of `layer_num=None` would make it much harder to handle serialization. I am currently using `layer_num` to extract the correct layer from `model` when deserializing, of course with this change it won't be possible. It also seems unlikely to be able to specify some layer identifier to automatically extract the layer from the model because the Bert example shows that you need to call a custom function (`get_input_embeddings`). We also don't want to save the whole model, it may be possible to save just the layer, but this then would be a separate copy of the layer unrelated to the layer in re-loaded model. One (not ideal) option is to extend the load API from `load_explainer('path', predictor)` to `load_explainer('path', predictor, **kwargs)` and require `layers` to be a `kwarg` specified by the user. Another option may be to provide a new `layer_extractor: Callable` `kwarg` so the user specifies the function used to extract the layer which we then can attempt to serialize, but this is also messy.
3. On a separate note, work on a PR proposal for automatically handling multi-output models (see https://github.com/SeldonIO/alibi/issues/343). The idea here is to come up with a reasonable solution for handling multi-output models natively. This would involve, as a minimum:
   - Figure out the `explain` API, e.g. do we add outputs to be explained as a list of `int`s, i.e. `outputs=[0,1]`, do we allow just one output to be explained at a time or multiple?
   - Figure out how to implement this internally, do we need to make an auxiliary model for every output to be explained or is there some other way?
   - Figure out how to change the returned attributions, currently `exp.attributions` is `List[np.ndarray]` supporting multi-input models (but not extending to multi-layer). Will need to figure out how this changes with multi-outputs.

**Comment by jklaise** [comment]
From a discussion:

2. The best scenario would be to find if there is an API to get layers by name, this would also solve the issue for explaining intermediate layers.
4. Related to multi-output models, the current implementation also assumes the outputs are 1 to 2-dimensional (for doing classification/regression), so currently the methods wouldn't work on higher-dimensional outputs. We would need to decide if we want to support this.
